instrumentation/spring/spring-webflux-5.0/auto/src/test/groovy/SpringWebfluxTest.groovy

```diff
@@ -173,7 +172,7 @@ class SpringWebfluxTest extends AgentTestRunner {
         }
         span(2) {
           operationName "tracedMethod"
-          childOf span(1)
+          childOf span(0)
```

Comment:
I think this change was wrong. Now the modeling isn't correct:

BEFORE:

```
netty handler
-> spring controller
---> traced method
```

AFTER
```
netty handler
-> spring controller
-> traced method
```

The traced method ends up a child of the netty span instead of the spring controller span. **This was the reason that I had captured the context at assembly time**

Reply:
This is not so obvious for me. If we run such application under profiler, then I expect that there will NOT be a single callstack with "netty -> spring controler -> traced method". Because traced method was actually executed AFTER returning from spring controller.

In addition I definitely expect that `[start:end]` of traced method IS NOT contained in `[start:end]` of spring controller. Currently it IS and I think this is a bug. Again, due to profiler/callstack argument above.

For these reasons I think that capturing context during assembly is wrong.

Reply:
I follow what you are saying but from a user point of view, I would consider that traced method to be a part of my controller call. I understand that it runs after the controller returns but my intent as the user would be that the traced method is part of the controller

Reply:
```
  @GetMapping("/foo-mono-from-callable/{id}")
  Mono<FooModel> getMonoFromCallable(@PathVariable("id") long id) {
    return Mono.fromCallable { return tracedMethod(id) }
  }

  @Trace()
  private FooModel tracedMethod(long id) {
    return new FooModel(id, "tracedMethod")
  }
```

I think this is an async modeling question but from a user point of view the model makes most sense to me to have `tracedMethod` as a child of `getMonoFromCallable`

Reply:
Yeah, that is kinda true. But there are examples of wrong traces in both cases: capture on assembly and capture on subscribe. I think the whole reactive story is currently very weak.

Reply:
This is an interesting example, and similar to:

```
  @GetMapping("/foo-callable/{id}")
  Callable<FooModel> getCallable(@PathVariable("id") long id) {
    return () -> { return tracedMethod(id) }
  }

  @Trace()
  private FooModel tracedMethod(long id) {
    return new FooModel(id, "tracedMethod")
  }
```

(https://docs.spring.io/spring/docs/current/spring-framework-reference/web.html#mvc-ann-async)

We wouldn't capture `traced method` under `controller` in this example either (since we capture context when passing the callable to the executor, and not when constructing the callable).

Reply:
What do you think of propagating the context from the controller to the return value when the return value represents an asynchronously executed operation? This seems similar to how we propagate the context across other async boundaries.

Reply:
I don't grok reactor very well but @trask's example is pretty easy to understand. By the way, do we handle the span correctly? If the span ends when `getCallable` returns, it seems broken even outside of just context propagation. This seems like a very special case.

My understanding here is that spring seems to offer a very specific execution model - the user "registers a task", but the actual execution happens by the framework. Because this task corresponds to a request, we want it to also correspond to a Span. This seems like it requires very specific handling - I think it's what you're saying Trask and I agree with it.

My feeling is this is a spring (or any framework that provides a similar task registration execution model) specific aspect, I don't think it should leak into Reactor. It does seem true that we can't really guarantee when subscription happens, maybe it happens in the wrong context. But the "reactive model" I think is - register many callbacks and then execute by calling subscribe, as opposed to completablefuture where you register callbacks while executing. But in user code I would expect the vast majority of usages to expect that subscription to be the parent - it is similar to our concept of an "invocation context" which we often remount, and it should apply to reactor as well, the subscriber is the invoker. If that doesn't fit with a separate abstraction, like webflux, we could handle that separately.

Unrelated to our instrumentation, but my understanding is that the challenge we have with reactive is they don't seem to define the concept of a scope but probably should - is the scope assembly, subscription, or something else, it's something that affects how we propagate and needs to be known. I personally use Dagger producers and never have context issues since it has a concept of a `ProductionComponent` which acts like a scope (similar to DI scopes). So all callbacks defined in that component run with the same context, the context of that scope. Works great, and no code required by the user as long as they follow the one requirement that the Dagger entrypoint needs to be executed in the request handler

`@Production` defines the executor used when invoking a component - so just set it to the request executor and problem solved :)
https://github.com/curioswitch/curiostack/blob/master/common/server/framework/src/main/java/org/curioswitch/common/server/framework/ServerModule.java#L250

Invocation looks like this. Yes if that invocation happened after being delegated to a threadpool it wouldn't work, but my framework just doesn't allow that (would throw guaranteed runtime exception), knowing how important context is.
https://github.com/curioswitch/curiostack/blob/master/cafe-map/server/src/main/java/org/curioswitch/cafemap/server/CafeMapService.java#L63

Unlike my framework, reactor does not have such a requirement - but I think we can still expect common usage by users to be that the stream is "scoped" to the subscription point.

Reply:
@iNikem @trask Just random update on this point, @kojilin came up with an interesting example which further seems to make context at subscription time, not assembly time, make sense

https://github.com/line/armeria/issues/3051#issuecomment-752413533

If contexting at assembly time, `cache` will generally cause disasters.